File: src/main/java/com/hyperboat/daily/y2024/m10/d21/LC910.java
package com.hyperboat.daily.y2024.m10.d21;

import java.util.Arrays;

public class LC910 {

  public int smallestRangeII(int[] nums, int k) {
    var n = nums.length;
    Arrays.sort(nums);
    var ans = nums[n - 1] - nums[0];
    for (var i = 0; i < n - 1; i++) {
      var x = nums[i];
      var y = nums[i + 1];
      var max = Math.max(nums[n - 1] - k, x + k);
      var min = Math.min(nums[0] + k, y - k);
      ans = Math.min(ans, max - min);
    }
    return ans;
  }
}

File: src/main/java/com/hyperboat/daily/y2024/m10/d21/LC910.kt
package com.hyperboat.daily.y2024.m10.d21

import java.util.*

internal class Solution {
    fun smallestRangeII(nums: IntArray, k: Int): Int {
        val n = nums.size
        Arrays.sort(nums)
        var ans = nums[n - 1] - nums[0]
        for (i in 0 until n - 1) {
            val x = nums[i]
            val y = nums[i + 1]
            val max = (nums[n - 1] - k).coerceAtLeast(x + k)
            val min = (nums[0] + k).coerceAtMost(y - k)
            ans = ans.coerceAtMost(max - min)
        }
        return ans
    }
}